### PR TITLE
List atomic types and make them accessible from python

### PR DESCRIPTION
**Pull request by astolman**
This does two things: 
1.  It adds a method to the C++ EntityTypeManager API to list the atomic type names
2.  Adds a cython wrapper for EntityTypeManager and an EntityType class in python to represent types. This makes it possible to list all atomic types of a katana.local.Graph. 

If this is accepted, I'll open a corresponding PR in the enterprise repo to add this functionality to `katana.distributed.Graph`.

For now, this only addresses atomic types. I think extending this to intersection types is a bigger problem that should be handled in another PR.

### CODE REVIEW

File: python/katana/cpp/libgalois/graphs/Graph.pxd
Comment (witchel) [inline python/katana/cpp/libgalois/graphs/Graph.pxd]: Should you also provide `GetEdgeTypeManager()`?

File: libsupport/include/katana/EntityTypeManager.h
Comment (witchel) [inline libsupport/include/katana/EntityTypeManager.h]: This looks fine, but bear in mind this is only atomic types.  A node might have type` {Message, Post}`, which (confusingly) is the intersection of `Message `and `Post`.

**Review by witchel — APPROVED** [review]
This looks fine to me, but I don't know if the real type-heads (for example @amberhassaan and @roshandathathri and @arthurp) object to listing atomic type names.

File: libsupport/include/katana/EntityTypeManager.h
Comment (astolman) [reply libsupport/include/katana/EntityTypeManager.h]: That's totally fine from my point of view. I'm adding this because types ultimately correspond to different feature representations for the ML stuff. How to handle intersection types for feature purposes can/will probably be handled in python eventually.

File: python/katana/cpp/libgalois/graphs/Graph.pxd
Comment (astolman) [reply python/katana/cpp/libgalois/graphs/Graph.pxd]: I don't really care about edge types right now for what I'm doing, but yes, it's probably a good idea.

Comment (arthurp) [reply python/katana/cpp/libgalois/graphs/Graph.pxd]: Please do.

File: python/katana/local/_graph.pyx
Comment (arthurp) [inline python/katana/local/_graph.pyx]: Let's not expose lists of strings. It would be much better to return a list of objects which represent the types. That way both check the type when they are passed back to us, and provide additional operations (like intersection) on types.

File: libsupport/include/katana/EntityTypeManager.h
Comment (arthurp) [reply libsupport/include/katana/EntityTypeManager.h]: I don't really think this is the API we want. I think we want something that returns a vector of the EntityTypeIDs. Then we can build a smart object around them at the python level. That object can get the names as needed.

File: python/katana/local/_graph.pyx
Comment (arthurp) [reply python/katana/local/_graph.pyx]: For now the object can just provide the name. But I think we will want it to do more eventually. It could be declared as:

```
cdef class EntityType:
  def __init__(self, type_manager, type_id):
    self._type_manager = type_manager
    self._type_id = type_id
  def __str__(self):
    return self._type_manager.GetAtomicTypeName(self.type_id)
```
I didn't test this, but it should be close to what you will want. You can then instantiate them based on a vector of IDs with a list comprehension.

File: libsupport/include/katana/EntityTypeManager.h
Comment (arthurp) [reply libsupport/include/katana/EntityTypeManager.h]: BTW, I would not right an intersection type like that because it is confusing. I would write `Message & Post`.

Comment (witchel) [reply libsupport/include/katana/EntityTypeManager.h]: I think you are right that `Message & Post` is a better model, though it still doesn't scream "intersection" at me.  I wrote `{Message, Post}` because types get passed to graph update as sets of strings.

File: python/katana/local/_graph.pyx
Comment (astolman) [reply python/katana/local/_graph.pyx]: I see your point, but I would like to balance it with a concern for maintaining a super simple python interface. Data scientists are sometimes not the best coders and I worry that adding too many abstractions would scare some of them away. I am basing my implementation roughly on DGL, which uses strings for type names in their HeteroGraph.

What if we implemented this EntityType class but still exported an API that lets users just deal in strings if that's what they're comfortable with? I would keep the node_type_names property but implement it to use the EntityType class.

Comment (arthurp) [reply python/katana/local/_graph.pyx]: I see your point, but I think we should push for better APIs to help data scientists improve. I think we can make the EntityType actually pretend to be a string in most cases, but still carry the extra information and have the extra properties. This might give us a best of both worlds.

Also I expected `node_type_names` or `node_types` (whatever we want to call it) to return EntityTypes. I'm not sure what else you were envisioning.

I would prefer we start with a strict API with a EntityType and then relax it as needed. It's much easier to add support for other types than remove support for unsafe code that may already have been written. One reason for EntityType, BTW, is that it could check whether the type is of the correct kind, node vs edge, when the type is used. Which would be quite useful for catching mistakes.

All this said, I can come back with a PR to make the API a bit stricter after you merge this if you don't think it's easy to do in this PR.

File: libsupport/include/katana/EntityTypeManager.h
Comment (arthurp) [reply libsupport/include/katana/EntityTypeManager.h]: re: `&` as intersection: If I write `a & b` (where `a` and `b` are say 64-bit integers) the set of 1-bits in the result will be exactly the intersection of the 1-bits in `a` and `b`. Bitwise and *is* intersection of small finite sets. Just as bitwise or is union.

I'm also happy to use `Message ∩ Post` if you prefer though. Unicode FTW.

Comment (arthurp) [reply libsupport/include/katana/EntityTypeManager.h]: Fan boy rant: Scala 3 supports type intersections using `&`. https://docs.scala-lang.org/scala3/book/types-intersection.html It's quite lovely.

File: python/katana/local/_graph.pyx
Comment (astolman) [reply python/katana/local/_graph.pyx]: Seems easy enough, and you make a compelling case. I'm working on it. I do just want to keep this concern for the lowest common denominator python user in mind though as we develop the API. For now, EntityTypes it is.

File: libsupport/include/katana/EntityTypeManager.h
Comment (amberhassaan) [reply libsupport/include/katana/EntityTypeManager.h]: I thought @aneeshdurg had a PR that added this API and I instead wanted an iterator over `std::string` that is build out of the iterators over the `map`.  See `boost::transform_iterator`. 

@arthurp : I see that folks are asking to iterate over `string` names, and I think an iterator is a good compromise, where we don't stick to building a vector but we also provide a cheap method of iteration over `strings`. By forcing folks to use `EntityTypeIDs` , we may be forcing duplication of not so straightforward code that would have been straightforward given the right API.

Comment (amberhassaan) [reply libsupport/include/katana/EntityTypeManager.h]: @astolman : please check with @aneeshdurg as he wrote the exact same code in a PR the other day (if I recall).

Comment (arthurp) [reply libsupport/include/katana/EntityTypeManager.h]: This seems orthogonal to me. Iterators vs vectors and names vs IDs are separate issues. I'm not worried about performance really, I'm worried about separation of concern. I don't think there will be any duplicated code. We can have a "list converter" similar to `GetEntityTypeIDs` (see below) which converts a contain of one type to another, to work around the lack of an idiomatic way to do iterator comprehensions in C++.

https://github.com/KatanaGraph/katana/blob/e8cfb244812db5bd561b2e8c4af37162433b4c21/libsupport/include/katana/EntityTypeManager.h#L291-L308

Side note: I would prefer a range to an interator.

I'm clearly in the minority in this debate though, so if strings are where people want it, so be it. I do reserve the right to laugh when weird bugs crop up or people have to start converting IDs to strings to call methods, though. ;-)

File: python/katana/local/_graph.pyx
Comment (arthurp) [reply python/katana/local/_graph.pyx]: You are right and I do have that in mind: https://katanagraph.atlassian.net/browse/KAT-628. It should be an epic. There are lots of bits involved. I'll change it. Feel free to add things to it if you think there are more potential confusions to be avoided.

My overall goal is to provide an API that is mostly familiar to a data-scientist, but also provides a bit more safety than they may be useful. Things like raising an exception if they try to index a node property with an edge index. But I still want it to integrate well with other code, so it will not be "too safe" since any special index types will still need to support conversion to int for indexing numpy arrays and the like. I'd love your input on all this.

File: libsupport/include/katana/EntityTypeManager.h
Comment (amberhassaan) [reply libsupport/include/katana/EntityTypeManager.h]: I agree that the two issues are orthagonal. Secondly, by `iterators` I mean `iterator ranges` for now, until we arrive at C++-20's ranges. 

TBH, I haven't spent time looking carefully at the various use cases of `EntityTypeManager's API`. But I have heard @witchel asking for the same, so I took the position of "the interface should provide it in an efficient yet generic manner". It could be that these folks are wrong and you are right and you will be laughing at us :).

File: python/katana/local/_graph.pyx
Comment (witchel) [reply python/katana/local/_graph.pyx]: I would welcome a meeting about Types and type management.  I like Arthur's idea of EntityTypes as an actual type that might have a string or a numeric representation.  If I got those from the Cypher engine, I'd be happy to process them and pass them to the partitioner.  Of course the C++ and python types don't need to be exactly 1:1, but the way Aurthur describes it, this is an opportunity to use the type system to our advantage for little performance cost.
What I don't like about the current interface is how clients have to understand the underlying representation to do things like iterate over all compound types.  Ranges are fine if our ID space is always guaranteed to be dense.  Is it?

**Comment by astolman** [comment]
I overhauled this PR per @arthurp's suggestions. EntityType is now a python class that represents the node/edge types. For now only atomic types are supported. I figure extending it to intersection types is a larger conversation, and I only need atomic types for now.

File: python/katana/local/_graph.pyx
Comment (arthurp) [inline python/katana/local/_graph.pyx:288]: NIT: I don't think you need the type here. I think it can infer it.

File: python/katana/local/entity_type.pyx
Comment (arthurp) [inline python/katana/local/entity_type.pyx]: NIT: This should be `make` not `Make`.

File: python/test/test_entity_type_manager.py
Comment (arthurp) [inline python/test/test_entity_type_manager.py:7]: ... I wonder why we included the `k` in the string version of this. Oh well. Not important.

File: python/katana/local/entity_type.pyx
Comment (arthurp) [inline python/katana/local/entity_type.pyx]: Is this needed to avoid issues with non-default constructable types? If so, document it in a comment. I know I haven't done that my self. Please call me out on that in the future so I fix it.

File: python/katana/local/entity_type.pyx
Comment (arthurp) [inline python/katana/local/entity_type.pyx]: The string type in python is spelled `str`.

File: python/katana/local/_graph.pyx
Comment (arthurp) [inline python/katana/local/_graph.pyx:298]: There is a convention of writing `list[T]` to give the element type, so this can be `list[EntityType]`.

**Review by arthurp — APPROVED** [review]
This is great. Just nits really. Thanks.

File: python/katana/local/entity_type.pyx
Comment (astolman) [reply python/katana/local/entity_type.pyx]: I don't know to be honest. I fought with this forever. Eventually this configuration caused it to pass the tests, so I'm afraid to change it.

Comment (arthurp) [reply python/katana/local/entity_type.pyx]: Could you just add a comment like? "This function seems to be needed to make the terrible confusing Cython compiler happy."

File: python/katana/local/_graph.pyx
Comment (astolman) [reply python/katana/local/_graph.pyx:288]: The compiler actually complains when I remove it about some kind of invalid conversion. It does seem necessary here.

File: python/katana/local/entity_type.pyx
Comment (astolman) [reply python/katana/local/entity_type.pyx]: Okay, I think that change might not have been what mollified the cython compiler. I'm changing it to just one function in the next commit

File: python/katana/local/_graph.pyx
Comment (ddn0) [reply python/katana/local/_graph.pyx:288]: This is an example where the convention of field accessors being snake_case is helpful because you know that you can take an address of a field, but if `GetNodeTypeManager` were to return a value rather than a reference, we'd have problems.

**Pull request merged**
